File: express-typescript/src/Book/Application/BookFinder.ts
import BookRepository from "../Domain/BookRepository";
import Book from "../Domain/Book";
import {inject, injectable} from "tsyringe";

@injectable()
export class BookFinder
{

    constructor( @inject("BookRepository") private repository: BookRepository)
    {
    }

    async run(title: string): Promise<Book | null>
    {
        return await this.repository.search(title);
    }
}

File: gin/src/Book/Application/BookFinder.go
package Application

import (
	"gin-ddd/src/Book/Domain"
)

type BookFinder struct {
	repository Domain.BookRepository
}

func NewBookFinder(repository Domain.BookRepository) *BookFinder {
	return &BookFinder{repository: repository}
}

func (bf BookFinder) Run(title string) Domain.Book {
	book, err := bf.repository.Search(title)
	if err != nil {
		panic(err.Error())
	}
	return book
}
